src/gleam/string.gleam

```diff
@@ -258,9 +316,51 @@ pub fn join(strings: List(String), with separator: String) -> String {
 /// pad_left("121", to: 2, with: ".") == "121"
 /// ```
 ///
-// pub fn pad_left(string: String, to size: Int, with: String) {}
+type PadSide {
+  Left
+  Right
+}
+
+fn pad_fill_help(result: String, chunk: String, chunk_length: Int, to_fill: Int) -> String {
+    let one = 1
+    case True {
+        True if one > to_fill->
+          result
+
+        True if to_fill > chunk_length ->
+          pad_fill_help(append(result, chunk), chunk, chunk_length, to_fill - chunk_length)
+
+        True ->
+          append(result, slice(chunk, 0, to_fill))
+    }
+}
+
+fn pad_help(pad: String, to_length: Int, with: String, pad_side: PadSide) -> String {
```

Comment:
I think we can improve the performance of this function.

1. Find the length of the string to pad
2. Check if the string is already long enough
3. Find the length of the string to pad with
4. Repeat the padding string by the number of times required to make the string the right length
5. Concat the binaries

Appending strings copies the strings, so we want to do this as few times as possible. If we want to build a string over multiple small parts it's best to do so using the iodata type as it is an efficient way of building strings.

Reply:
oh yeah good call! build then concat rather than append.

This implementation is slightly more complicated than that only because it allows for a multi-grapheme string to pad with.

Rather than 
  - determine how long the pad needs to be
  - determine how many repetitions of the prefix to be at least that long
  - repeat the prefix that many times
  - cut prefix to correct length

it:
  - determine how long the pad needs to be
  - loop until it would be too long
  - cut the last chunk to the correct length
  - join

I'm not partial to either though and am happy to implement it either way.

This function kind of has to support multi-grapheme string to pad with though otherwise it would need to return an option if a non single-grapheme string was passed in.

Reply:
Getting the length of a string is linear so we shouldn't do that in the resursive helper function. If we know the length of the original and the padding chunk we can calculate the number of chunks required in advance, can't we?

Reply:
for sure, the lengths are only calculated once at the top of `pad_help` outside of any recursion

Reply:
here's some pseudo code for possibly a nicer way to implement this:
```
if total_length > current_length ->
  partial_chunk_length = total_length % chunk
  [repeat(chunk, fill_length / chunk_length), slice(chunk, 0, partial_chunk_length), input_string ]
  |> concat
```